lib/semantics/resolve-names.cc

```diff
@@ -5027,7 +4996,77 @@ static bool NeedsExplicitType(const Symbol &symbol) {
   }
 }
 
-void ResolveNamesVisitor::Post(const parser::SpecificationPart &) {
+bool ResolveNamesVisitor::Pre(const parser::SpecificationPart &x) {
+  Walk(std::get<0>(x.t));
+  Walk(std::get<1>(x.t));
+  Walk(std::get<2>(x.t));
+  Walk(std::get<3>(x.t));
+  const std::list<parser::DeclarationConstruct> &decls{std::get<4>(x.t)};
+  for (const auto &decl : decls) {
+    if (const auto *spec{
+            std::get_if<parser::SpecificationConstruct>(&decl.u)}) {
+      PreSpecificationConstruct(*spec);
+    }
+  }
+  Walk(decls);
+  FinishSpecificationPart();
+  return false;
+}
+
+// Initial processing on specification constructs, before visiting them.
+void ResolveNamesVisitor::PreSpecificationConstruct(
+    const parser::SpecificationConstruct &spec) {
+  std::visit(
+      common::visitors{
+          [&](const Indirection<parser::DerivedTypeDef> &y) {},
+          [&](const parser::Statement<Indirection<parser::GenericStmt>> &y) {
+            CreateGeneric(std::get<parser::GenericSpec>(y.statement.value().t));
+          },
+          [&](const Indirection<parser::InterfaceBlock> &y) {
+            const auto &stmt{std::get<parser::Statement<parser::InterfaceStmt>>(
+                y.value().t)};
+            const auto *spec{std::get_if<std::optional<parser::GenericSpec>>(
+                &stmt.statement.u)};
+            if (spec && spec->has_value()) {
+              CreateGeneric(**spec);
+            }
+          },
+          [&](const auto &) {},
+      },
+      spec.u);
+}
+
+void ResolveNamesVisitor::CreateGeneric(const parser::GenericSpec &x) {
+  auto info{GenericSpecInfo{x}};
+  const SourceName &symbolName{info.symbolName()};
+  if (IsLogicalConstant(context(), symbolName)) {
+    Say(symbolName,
+        "Logical constant '%s' may not be used as a defined operator"_err_en_US);
+    return;
+  }
+  GenericDetails genericDetails;
+  if (Symbol * existing{currScope().FindSymbol(symbolName)}) {
+    if (existing->has<GenericDetails>()) {
+      info.Resolve(existing);
+      return;  // already have generic, add to it
+    }
+    Symbol &ultimate{existing->GetUltimate()};
+    if (ultimate.has<GenericDetails>()) {
+      genericDetails.AddSpecificProcsFrom(ultimate);
```

Comment:
Should potential derived type/subroutine with the same name also be copied ?
Currently the following program gives error `error: 'foo' is not a derived type` for `type(foo) something`

```
module m1
  interface foo
    procedure :: f
  end interface
  type foo
    integer mbar
  end type
  contains
    function f()
    end function
end module

module m2
  use m1
  interface foo
    procedure :: f2
  end interface
  contains
    function f2(x)
    end function
end module

use m2
type(foo) something
end
```
It compiles fine if `foo` is not a generic in `m1`. NAG and gfortran and pgfortran compile this ok.

Reply:
The comment may also apply to other places that are "copying" genericDetails, where it seems to me it would make sens to also bring the type/subroutine e.g:
https://github.com/flang-compiler/f18/blob/1c32a289b53d787c0033ed45033db86afde33458/lib/semantics/resolve-names.cc#L1987
https://github.com/flang-compiler/f18/blob/1c32a289b53d787c0033ed45033db86afde33458/lib/semantics/resolve-names.cc#L1999

Reply:
Thanks, Jean, you are right, that should work. I've pushed a fix for it.